core/services/llo/onchain_channel_definition_cache.go

```diff
@@ -0,0 +1,260 @@
+				} else {
+					c.lggr.Debugw("No new channel definitions", "nLogs", 0, "definitionsBlockNum", c.definitionsBlockNum)
+				}
+			}
+		}
+	}
+}
+
+func (c *channelDefinitionCache) fetchFromChain() (nLogs int, err error) {
+	// TODO: Pass context
+	// https://smartcontract-it.atlassian.net/browse/MERC-3653
+	latest, err := c.lp.LatestBlock()
+	if errors.Is(err, sql.ErrNoRows) {
+		return 0, nil
+	} else if err != nil {
+		return 0, err
+	}
+	toBlock := latest.BlockNumber
+
+	fromBlock := c.definitionsBlockNum
+
+	if toBlock <= fromBlock {
+		return 0, nil
+	}
+
+	ctx, cancel := services.StopChan(c.chStop).NewCtx()
+	defer cancel()
+	// NOTE: We assume that log poller returns logs in ascending order chronologically
+	logs, err := c.lp.LogsWithSigs(fromBlock, toBlock, allTopics, c.addr, pg.WithParentCtx(ctx))
+	if err != nil {
+		// TODO: retry?
+		// https://smartcontract-it.atlassian.net/browse/MERC-3653
+		return 0, err
+	}
+	for _, log := range logs {
+		if err = c.applyLog(log); err != nil {
+			return 0, err
+		}
+	}
+
+	// Use context.Background() here because we want to try to save even if we
+	// are closing
+	if err = c.orm.StoreChannelDefinitions(context.Background(), c.addr, c.Definitions(), toBlock); err != nil {
+		return 0, err
+	}
+
+	c.definitionsBlockNum = toBlock
+
+	return len(logs), nil
+}
+
+func (c *channelDefinitionCache) applyLog(log logpoller.Log) error {
+	switch log.EventSig {
+	case topicNewChannelDefinition:
+		unpacked := new(channel_config_store.ChannelConfigStoreNewChannelDefinition)
+
+		err := channelConfigStoreABI.UnpackIntoInterface(unpacked, "NewChannelDefinition", log.Data)
+		if err != nil {
+			return fmt.Errorf("failed to unpack log data: %w", err)
+		}
+
+		c.applyNewChannelDefinition(unpacked)
+	case topicChannelDefinitionRemoved:
+		unpacked := new(channel_config_store.ChannelConfigStoreChannelDefinitionRemoved)
+
+		err := channelConfigStoreABI.UnpackIntoInterface(unpacked, "ChannelDefinitionRemoved", log.Data)
+		if err != nil {
+			return fmt.Errorf("failed to unpack log data: %w", err)
+		}
+
+		c.applyChannelDefinitionRemoved(unpacked)
+	default:
+		c.lggr.Errorw("Unexpected log topic", "topic", log.EventSig.Hex())
+	}
+	return nil
+}
+
+func (c *channelDefinitionCache) applyNewChannelDefinition(log *channel_config_store.ChannelConfigStoreNewChannelDefinition) {
+	rf := DecodeReportFormat(log.ChannelDefinition.ReportFormat)
+	streamIDs := make([]commontypes.StreamID, len(log.ChannelDefinition.StreamIDs))
+	copy(streamIDs, log.ChannelDefinition.StreamIDs)
+	c.definitionsMu.Lock()
+	defer c.definitionsMu.Unlock()
+	c.definitions[log.ChannelId] = commontypes.ChannelDefinition{
+		ReportFormat:  rf,
+		ChainSelector: log.ChannelDefinition.ChainSelector,
+		StreamIDs:     streamIDs,
+	}
+}
+
+func DecodeReportFormat(onchainRF [8]byte) commontypes.LLOReportFormat {
+	n := bytes.Index(onchainRF[:], []byte{0})
```

Comment:
this seems delicate. what happens if the first byte is `0`

what if you want to support a chain with a long string name that doesn't naturally fit into 8 bytes?

maybe decode into an integer and use that integer as an enum to get the string value

Reply:
First byte being `0` would make this the empty string, reportformat is a human-readable string.

This is an artifact of the datatype chosen on ethereum. `[8]byte` was chosen to hold it. Not sure why, I will ask @Fletch153 .

Reply:
Will change to enum, see: https://smartcontract-it.atlassian.net/browse/MERC-3662

Reply:
https://github.com/smartcontractkit/chainlink-common/pull/352